pairing/bn256/gfp.go

```diff
@@ -96,3 +129,30 @@ func (e *gfP) BigInt() *big.Int {
 
 func montEncode(c, a *gfP) { gfpMul(c, a, r2) }
 func montDecode(c, a *gfP) { gfpMul(c, a, &gfP{1}) }
+
+func sign0(e *gfP) int {
+	x := &gfP{}
+	montDecode(x, e)
+	for w := 3; w >= 0; w-- {
+		if x[w] > pMinus1Over2[w] {
+			return 1
+		} else if x[w] < pMinus1Over2[w] {
+			return -1
+		}
+	}
+	return 1
+}
+
+func legendre(e *gfP) int {
+	f := &gfP{}
+	// Since p = 4k+3, then e^(2k+1) is the Legendre symbol of e.
+	f.exp(e, pMinus1Over2)
+
+	montDecode(f, f)
+
+	if *f != [4]uint64{} {
```

Comment:
I'd have put the degenerate case first for clarity, though perhaps there is a negligible performance benefit in the average case in its current form

Reply:
Let's keep these files in sync with Cloudflare's ones.